lighthouse-core/audits/preload-lcp-image.js

```diff
@@ -0,0 +1,167 @@
+  title: 'Preload Largest Contentful Paint image',
+  /** Description of a lighthouse audit that tells a user to preload an image in order to improve their LCP time.  */
+  description: 'Preload the image used by ' +
+    'the LCP element in order to improve your LCP time.',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class PreloadLCPImageAudit extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'preload-lcp-image',
+      title: str_(UIStrings.title),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['traces', 'devtoolsLogs', 'URL', 'TraceElements', 'ImageElements'],
+      scoreDisplayMode: Audit.SCORING_MODES.NUMERIC,
+    };
+  }
+
+  /**
+   *
+   * @param {LH.Artifacts.NetworkRequest} request
+   * @param {LH.Artifacts.NetworkRequest} mainResource
+   * @param {Array<LH.Gatherer.Simulation.GraphNode>} initiatorPath
+   * @param {string} lcpImageSource
+   * @return {boolean}
+   */
+  static shouldPreloadRequest(request, mainResource, initiatorPath, lcpImageSource) {
+    const mainResourceDepth = mainResource.redirects ? mainResource.redirects.length : 0;
+
+    // If it's not the request for the LCP image, don't recommend it.
+    if (request.url !== lcpImageSource) return false;
+    // If it's already preloaded, no need to recommend it.
+    if (request.isLinkPreload) return false;
+    // It's not a request loaded over the network, don't recommend it.
+    if (URL.NON_NETWORK_PROTOCOLS.includes(request.protocol)) return false;
+    // It's already discoverable from the main document, don't recommend it.
+    if (initiatorPath.length <= mainResourceDepth) return false;
+    // Finally, return whether or not it belongs to the main frame
+    return request.frameId === mainResource.frameId;
+  }
+
+  /**
+   * @param {LH.Artifacts.NetworkRequest} mainResource
+   * @param {LH.Gatherer.Simulation.GraphNode} graph
+   * @param {LH.Artifacts.TraceElement|undefined} lcpElement
+   * @param {Array<LH.Artifacts.ImageElement>} imageElements
+   * @return {string|undefined}
+   */
+  static getURLToPreload(mainResource, graph, lcpElement, imageElements) {
+    if (!lcpElement) return undefined;
+
+    const lcpImageElement = imageElements.find(elem => {
+      return elem.devtoolsNodePath === lcpElement.devtoolsNodePath;
+    });
+
+    if (!lcpImageElement) return undefined;
+    let lcpUrl;
+    graph.traverse((node, traversalPath) => {
+      if (node.type !== 'network') return;
+      const record = node.record;
+      const imageSource = lcpImageElement.src;
+      // Don't include the node itself or any CPU nodes in the initiatorPath
+      const path = traversalPath.slice(1).filter(initiator => initiator.type === 'network');
+
+      // eslint-disable-next-line max-len
+      if (!PreloadLCPImageAudit.shouldPreloadRequest(record, mainResource, path, imageSource)) return;
```

Comment:
want to move the `record.url` check into the `traverse` itself? we're essentially *only* trying to find the LCP node here so it's weird that we do all this other work for traversal path and whatnot for all nodes 

(for example it would make sense to me if instead we had a separate `findLcpNode` function that just returned the node and then we called `shouldPreloadRequest` exactly once :D)

Reply:
I see what you mean. The current way I went it about it seems really unnecessary now that I look at it again haha.